sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1142,17 +1230,15 @@ def __init__(self, state, provision_info):
     self.data_server.start()
     self.control_server.start()
 
-  def data_api_service_descriptor(self):
-    return endpoints_pb2.ApiServiceDescriptor(
-        url='localhost:%s' % self.data_port)
-
-  def state_api_service_descriptor(self):
-    return endpoints_pb2.ApiServiceDescriptor(
-        url='localhost:%s' % self.state_port)
+  @classmethod
+  def get_instance(cls, state, provision_info):
```

Comment:
This method and `release_instance` should do the reference counting and decision whether to "actually" close and deallocate the server, not the callers. That keeps it more local and easier to reason about.